### ISSUE TITLE
Unsafe out= keyword argument with tensors sharing storage

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#5272]
The current `out=` behavior is unsafe. For example, if you do `a_3x3_tensor.sum(dim=1, out=y[1, 1:10])` it will try to `THTensor_(resize)` the view and in turn write to incorrect locations of `y`. 

Instead, if `out` is sharing storage, we should accept if only it 1) has correct size and 2) the range is not overlapping with the source tensors.

cc @zou3519 @soumith

**Comment by gchanan** [comment]
See also https://github.com/pytorch/pytorch/issues/5272.

**Comment by zhangguanheng66** [comment]
@gchanan I guess you mean this one https://github.com/pytorch/pytorch/issues/28576

**Comment by gchanan** [comment]
No, I mean the other one.

**Comment by mruberry** [comment]
@zou3519 Was this issue resolved?

**Comment by zou3519** [comment]
Not resolved:
```

In [24]: y = torch.zeros(5, 5)
    ...: tensor = torch.ones(3, 3)
    ...: torch.sum(tensor, dim=1, out=y[1:3, 1])
    ...: y
Out[24]:
tensor([[0., 0., 0., 0., 0.],
        [0., 3., 3., 3., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]])
```
The problem is that although we specified a column of three values for y, torch.sum ended up writing to values outside the ones we specified for y! This can be concerning behavior for users; the real problem here is that our out= resize semantics aren't very sane.

**Comment by mruberry** [comment]
Thanks for clarifying. Linking this with https://github.com/pytorch/pytorch/issues/41027, which I think would address this issue.
